wcfsetup/install/files/lib/data/user/UserProfileAction.class.php

```diff
@@ -609,6 +614,83 @@ public function validateUploadCoverPhoto()
         $uploadHandler->validateFiles(new UserCoverPhotoUploadFileValidationStrategy());
     }
 
+    /**
+     * Sets an avatar for a given user.
+     *
+     * @throws UserInputException If none or more than one user is given.
+     * @throws \InvalidArgumentException If the given avatar is not an image or is incorrect sized.
+     * @throws \RuntimeException If the avatar could not be saved for any reasons.
+     */
+    public function setAvatar(): array
+    {
+        $user = $this->getSingleObject();
+
+        $imageData = \getimagesize($this->parameters['fileLocation']);
+
+        if (!$imageData) {
+            throw new \InvalidArgumentException("The given file is not an image.");
+        }
+
+        if ($imageData[0] != UserAvatar::AVATAR_SIZE || $imageData[1] != UserAvatar::AVATAR_SIZE) {
+            throw new \InvalidArgumentException(
+                "The given avatar has not the size of " . UserAvatar::AVATAR_SIZE . "×" . UserAvatar::AVATAR_SIZE . "."
+            );
+        }
+
+        $data = [
+            'avatarName' => $this->parameters['filename'] ?? 'avatar',
+            'avatarExtension' => $this->parameters['extension'] ?? ImageUtil::getExtensionByMimeType($imageData['mime']),
+            'width' => $imageData[0],
+            'height' => $imageData[1],
+            'userID' => $user->userID,
+            'fileHash' => \sha1_file($this->parameters['fileLocation']),
+        ];
+
+        // create avatar
+        $avatar = UserAvatarEditor::create($data);
+
+        // check avatar directory
+        // and create subdirectory if necessary
+        $dir = \dirname($avatar->getLocation(null, false));
+        if (!@\file_exists($dir)) {
```

Comment:
I don't see how the `file_exists` could fail. Remove the `@` and then put this into the `try` for the moving logic.